content/en/blog/_posts/2021-09-10-kubernetes-in-kubernetes-and-pxe-bootable-servers-farm.md

```diff
@@ -0,0 +1,204 @@
+
+* [**Dockerfile**](https://github.com/kvaps/kubefarm/blob/master/build/ltsp/Dockerfile)
+
+What's going on here? First, we take a regular Ubuntu 20.04 and install all the packages we need. First of all we install the **kernel**, **lvm**, **systemd**, **ssh**. In general, everything that you want to see on the final node should be described here. Here we also install **docker** with **kubelet** and **kubeadm**, which are used to join the node to the cluster.
+
+And then we make an additional configuration. In the last stage, we simply install **tftp** with **nginx** (which serves our image to clients), **grub** (bootloader). Then root of the previous stages copied into the final image and generate squashed image from it. That is, in fact, we get a docker image, which has both the server and the boot image for our nodes. At the same time, by changing the Dockerfile, it can be easily updated.
+
+### Webhooks and API aggregation layer
+
+I want to pay special attention to the problem of webhooks and aggregation layer. In general, webhooks is a Kubernetes feature that allows you to respond to the creation or modification of any resources. Thus, you can add a handler so that when resources are applied, Kubernetes must send request to some pod and check if configuration of this resource is correct, or make additional changes to it.
+
+But the point is, in order for the webhooks to work, the apiserver must have direct access to the cluster for which it is running. And if it is started in a separate cluster, like our case, or even separately from the any cluster, then Konnectivity service can help us here. Konnectivity is one of the optional but officially supported Kubernetes components.
+
+Let's take cluster of four nodes for example, each of them is running a **kubelet** and we have other Kubernetes components running outside: **apiserver**, **scheduller** and **controller-manager**. By default, all these components interact with the apiserver directly - this is the most known part of the Kubernetes logic. But in fact, there is also a reverse connections. For example, when you want to view the logs or run a **kubectl exec** command, apiserver establishes a connection to the specific kubelet independently:
+
+![](/images/blog/2021-09-10-kubernetes-in-kubernetes-and-pxe-bootable-servers-farm/konnectivity01.jpg)
+
+But the problem is that if we have a webhook, then it is usually runs as a standard pod with a service in our cluster. And when apiserver tries to reach it, it will fail because it will try to access an in-cluster service named **webhook.namespace.svc** being outside of the cluster where it is actually running:
```

Comment:
```suggestion
But the problem is that if we have a webhook, then it usually runs as a standard pod with a service in our cluster. And when apiserver tries to reach it, it will fail because it will try to access an in-cluster service named **webhook.namespace.svc** being outside of the cluster where it is actually running:
```